src/Compilers/CSharp/Portable/Syntax/SyntaxTreeDiagnosticEnumerator.cs

```diff
@@ -11,136 +13,121 @@ namespace Microsoft.CodeAnalysis.CSharp
-            {
-                _syntaxTree = syntaxTree;
-                _stack = new NodeIterationStack(DefaultStackCapacity);
-                _stack.PushNodeOrToken(node);
-            }
-            else
-            {
-                _stack = new NodeIterationStack();
-            }
-        }
+            if (!root.ContainsDiagnostics)
+                yield break;
 
-        /// <summary>
-        /// Moves the enumerator to the next diagnostic instance in the diagnostic list.
-        /// </summary>
-        /// <returns>Returns true if enumerator moved to the next diagnostic, false if the
-        /// enumerator was at the end of the diagnostic list.</returns>
-        public bool MoveNext()
-        {
-            while (_stack.Any())
+            using var stack = new NodeIterationStack(DefaultStackCapacity);
+            stack.PushNodeOrToken(root);
+
+            while (stack.Any())
             {
-                var diagIndex = _stack.Top.DiagnosticIndex;
-                var node = _stack.Top.Node;
-                var diags = node.GetDiagnostics();
-                if (diagIndex < diags.Length - 1)
-                {
-                    diagIndex++;
-                    var sdi = (SyntaxDiagnosticInfo)diags[diagIndex];
+                var node = stack.Top.Node;
 
-                    //for tokens, we've already seen leading trivia on the stack, so we have to roll back
-                    //for nodes, we have yet to see the leading trivia
-                    int leadingWidthAlreadyCounted = node.IsToken ? node.GetLeadingTriviaWidth() : 0;
+                if (!stack.Top.ProcessedDiagnostics)
+                {
+                    foreach (SyntaxDiagnosticInfo sdi in node.GetDiagnostics())
+                    {
+                        // For tokens, we've already seen leading trivia (as we push leading/trailing trivia explicit as
+                        // green nodes to process on the stack), so we have to roll back.  For nodes, we have yet to see the
+                        // leading trivia, so those don't need an adjustment.
+                        int leadingWidthAlreadyCounted = node.IsToken ? node.GetLeadingTriviaWidth() : 0;
 
-                    // don't produce locations outside of tree span
-                    Debug.Assert(_syntaxTree is object);
-                    var length = _syntaxTree.GetRoot().FullSpan.Length;
-                    var spanStart = Math.Min(_position - leadingWidthAlreadyCounted + sdi.Offset, length);
-                    var spanWidth = Math.Min(spanStart + sdi.Width, length) - spanStart;
+                        // don't produce locations outside of tree span
+                        var length = syntaxTree.GetRoot().FullSpan.Length;
+                        var spanStart = Math.Min(position - leadingWidthAlreadyCounted + sdi.Offset, length);
+                        var spanWidth = Math.Min(spanStart + sdi.Width, length) - spanStart;
```

Comment:
i'm doing an investigation into offsets, and makign them saner.  this was the part of the code that consumes them and does the most math on them.  so i wanted to make this as easy to understand manipulate for an upcomign PR.